### DISCUSSION TITLE
Add `light.move`

### DISCUSSION THREAD

**Discussion by agoode** [home-assistant/architecture#1094] [Entity Models]
Many light devices support a "start/stop" "raise/lower" command, which is used to smoothly change light levels based on physical remote press/release actions. This is important for two reasons:
- Sending start/stop commands minimizes radio transmissions while providing smooth dimming when controlled with buttons
- When lights are grouped, a move command will correctly change the light levels independently and by the same amount (all lights will change at the same rate for the same time, keeping their relative differences).

This kind of command is supported natively by Matter, Zigbee, Lutron, and probably others. (Hue has something close, but it doesn't quite work properly https://developers.meethue.com/forum/t/surprising-behavior-with-dimming-delta-on-grouped-light/7031).

In Matter and Zigbee, this is the `Move` command. Lutron calls it "Start Raising" and "Start Lowering" in the old integration protocol: https://assets.lutron.com/a/documents/040249.pdf. (There is probably a LEAP version, but it is undiscovered as yet.)

Right now, Home Assistant cannot implement this smooth level change. Workarounds exist: the behavior may be emulated by sending multiple small step commands, but these may exceed Zigbee's low multicast rate and does not work correctly with groups (https://github.com/home-assistant/core/issues/118009).

I think there are a few related light improvements worth investigating, all somewhat related to exposing more of the underlying light's native functionality. This one (`light.move`), https://github.com/home-assistant/architecture/discussions/537 (`light.adjust`), and also some better group handling. I'm not sure the best way to make progress on all these, but I will try to keep this one to a small scope.

**Comment by elupus** [comment]
Knx can be added to that list.

**Comment by InovelliUSA** [comment]
Have been wanting support for this for some time. To add to the conversation, Z-Wave also has this functionality via the MultiLevelSwitch command class. The commands are StartLevelChange (up or down) and StopLevelChange. It would be nice if all protocols that support this functionality (Matter, Z-Wave, Hue Bridge, Zigbee, etc.) could be supported which would allow for cross protocol integrations. Looks like Z-Wave JS supports this command currently. 
![image](https://github.com/user-attachments/assets/2dc799b3-2371-4a75-9200-77bad1f6b7ea)

**Comment by tringenbach** [comment]
The Bond integration has this feature or similar too https://github.com/home-assistant/core/blob/dev/homeassistant/components/bond/light.py#L196

But it's deprecated, and the button approach is awkward.  And either way it's non-standard. As a user, I want this part of the generic light integration, so that my automations don't have to care what kind of light it is.

There is https://www.home-assistant.io/integrations/light/#action-lightturn_on `transition` already, which is close, but you have to specify how much you're increasing / decreasing ahead of time, and there's no "stop".

**Comment by nohat** [comment]
When a dimming button on a remote is held, the underlying protocols can already begin a continuous brightness change and stop it on release: the Zigbee Cluster Library's Level cluster specifies Move and Stop, Z-Wave's Multilevel Switch command class has Start Level Change and Stop Level Change, and Hue's CLIP v2 API provides `dimming_delta` with an explicit stop action. I added the equivalent commands to Tasmota in 2021 ([arendst/Tasmota#11269](https://github.com/arendst/Tasmota/pull/11269)). The capability has been present in the protocols and in much of the firmware for years; what has been missing is what this proposal describes — a uniform way for Home Assistant to express it.

To put evidence under that case I built [dynamic_dimming](https://github.com/nohat/dynamic_dimming), a HACS integration that provides the service surface and dispatches each call to the device's own protocol where one exists, falling back to stepped simulation where it does not:

- `dynamic_dimming.move` / `stop` / `step` take an entity and a direction; nothing about the transport reaches the caller.
- `capability.classify()` sorts each light into `native`, `simulated`, or `unsupported`.
- the Zigbee2MQTT backend publishes the Level cluster's `brightness_move` / `brightness_move: stop`; the Tasmota backend publishes the `Dimmer >` / `<` / `!` from #11269. Each backend is a small class against a common interface (`claims`, `async_move`, `async_stop`, `async_step`), so ZHA or Z-Wave JS is another file rather than a redesign.

It is early (v0.1b), but it runs the interaction end to end on real hardware, and it carries a device-report form so behavior can be recorded across setups rather than argued in the abstract. I'd rather contribute that to the proposal you've framed here than open a competing one. Let me know whether that would be useful, or if your current thinking about the interface is different from what you first proposed.

**Comment by abmantis** [comment]
What are the `light.move` service parameters? How to stop?
I wonder if `light.brightness_up`/`light.brightness_down`/`light.brightness_stop` would be better? It also aligns with `cover.open_cover`/`cover.close_cover`.

**Reply by farmio** [reply]
You might not want to change brightness, but color_temperature / color. Or both.

**Reply by abmantis** [reply]
> You might not want to change brightness, but color_temperature / color. Or both.

That is why I suggested `brightness_up`, etc, because later we could add `color_temperature_up`.

With a single `light.move` service, what would be the args to allow that?

**Reply by farmio** [reply]
Oops, I confused this with the `light.adjust' discussion 😬 Sorry. 

But still, for a `light.move` we could have step attributes for brightness and color temperature. Like 20% brighter, 800K colder in a single command.

**Reply by abmantis** [reply]
> Oops, I confused this with the `light.adjust' discussion 😬 Sorry.
> 
> But still, for a `light.move` we could have step attributes for brightness and color temperature. Like 20% brighter, 800K colder in a single command.

I am not 100% sure, but the proposal seems to be about "raise until a stop is sent" and not about "raise by 20%". But that needs to be clarified on the proposal I guess.

**Reply by erocm123** [reply]
The Z-wave implementation supports:

| Parameter | Type | Description |
|-----------|------|-------------|
| `ignoreStartLevel` | Boolean | If `true`, starts from the current level. If `false`, uses `startLevel`. |
| `startLevel` | Integer (0-99) | Starting brightness level (used only when `ignoreStartLevel` is `false`). |
| `duration` | Duration *(optional)* | Transition duration, if supported by the device. |
| `secondarySwitch` | Boolean *(optional)* | Indicates the command originated from a secondary rocker/button. Rarely used. |

Zigbee kind of supports 'duration' based on how high or low the negative or positive value is:

| Parameter | Type | Description |
|-----------|------|-------------|
| `brightness_move` | Integer | Signed move rate. Positive values brighten, negative values dim, and `0` stops the movement. |

I think the obvious overlap is the "duration" which you could call "rate" or "speed". So, at minimum I think it should support direction and rate:

## Recommended Home Assistant API

Home Assistant should expose one protocol-neutral parameter:

```yaml
action: light.move
target:
  entity_id: light.kitchen
data:
  direction: up
  rate (or speed?): 40
```

Rate (or speed) can be:
> **How long should it take to go from 0% to 100%?**

That's exactly how Z-Wave already models it, and it's intuitive.

For example:

```yaml
action: light.move
target:
  entity_id: light.kitchen
data:
  direction: up
  rate: 10
```

where `rate` is defined as:

> **The number of seconds required to traverse the full brightness range (0–100%).**

### Then each integration converts it:

| HA `rate` (seconds) | Z-Wave | Zigbee |
|--------------------:|---------|---------|
| 1 | `duration = 1 s` | `brightness_move = 254` |
| 2 | `duration = 2 s` | `brightness_move = 127` |
| 5 | `duration = 5 s` | `brightness_move = 51` |
| 10 | `duration = 10 s` | `brightness_move = 25` |
| 20 | `duration = 20 s` | `brightness_move = 13` |

The Zigbee integration would simply compute:

```text
brightness_move = round(254 / rate)
```

For example:

```yaml
rate: 10
```

would become:

```text
brightness_move = 25
```

The 0.4 level/second rounding error is negligible and well within the precision of the protocol.

## Why this API is better

A user thinks:

> "I want this to dim over about 10 seconds."

They do **not** think:

> "I want 25.4 brightness levels per second."


Anyway, just my input to consider.
